swarm/storage/mru/signedupdate.go

```diff
@@ -0,0 +1,198 @@
+func (r *SignedResourceUpdate) Sign(signer Signer) error {
+
+	r.binaryData = nil           //invalidate serialized data
+	digest, err := r.getDigest() // computes digest and serializes into .binaryData
+	if err != nil {
+		return err
+	}
+
+	signature, err := signer.Sign(digest)
+	if err != nil {
+		return err
+	}
+
+	// Although the Signer interface returns the public address of the signer,
+	// recover it from the signature to see if they match
+	ownerAddress, err := getOwner(digest, signature)
+	if err != nil {
+		return NewError(ErrInvalidSignature, "Error verifying signature")
+	}
+
+	if ownerAddress != signer.Address() { // sanity check to make sure the Signer is declaring the same address used to sign!
+		return NewError(ErrInvalidSignature, "Signer address does not match ownerAddr")
+	}
+
+	r.signature = &signature
+	r.updateAddr = r.UpdateAddr()
+	return nil
+}
+
+// create an update chunk.
+func (r *SignedResourceUpdate) toChunk() (*storage.Chunk, error) {
+
+	// Check that the update is signed and serialized
+	// For efficiency, data is serialized during signature and cached in
+	// the binaryData field when computing the signature digest in .getDigest()
+	if r.signature == nil || r.binaryData == nil {
+		return nil, NewError(ErrInvalidSignature, "newUpdateChunk called without a valid signature or payload data. Call .Sign() first.")
+	}
+
+	chunk := storage.NewChunk(r.updateAddr, nil)
+	resourceUpdateLength := r.resourceUpdate.binaryLength()
+	chunk.SData = r.binaryData
+
+	// signature is the last item in the chunk data
+	copy(chunk.SData[resourceUpdateLength:], r.signature[:])
+
+	chunk.Size = int64(len(chunk.SData))
+	return chunk, nil
+}
+
+// fromChunk populates this structure from chunk data
+func (r *SignedResourceUpdate) fromChunk(updateAddr storage.Address, chunkdata []byte) error {
+	// for update chunk layout see SignedResourceUpdate definition
+
+	//deserialize the resource update portion
+	if err := r.resourceUpdate.binaryGet(chunkdata); err != nil {
+		return err
+	}
+
+	// Extract the signature
+	var signature *Signature
+	cursor := r.resourceUpdate.binaryLength()
+	sigdata := chunkdata[cursor : cursor+signatureLength]
+	if len(sigdata) > 0 {
+		signature = &Signature{}
+		copy(signature[:], sigdata)
+	}
+
+	// check that the lookup information contained in the chunk matches the updateAddr (chunk search key)
+	// that was used to retrieve this chunk
+	// if this validation fails, someone forged a chunk.
+	if !bytes.Equal(updateAddr, r.updateHeader.UpdateAddr()) {
+		return NewError(ErrInvalidSignature, "period,version,rootAddr contained in update chunk do not match updateAddr")
+	}
+
+	r.signature = signature
+	r.updateAddr = updateAddr
+	r.binaryData = chunkdata
+
+	// Verify signatures and that the signer actually owns the resource
+	// If it fails, it means either the signature is not valid, data is corrupted
+	// or someone is trying to update someone else's resource.
+	if err := r.Verify(); err != nil {
```

Comment:
Should this check be here? This is called by `handler.Validator` and `handler.updateIndex`, and the latter doesn't really need the check, no?

Reply:
I could remove it and add `Verify()` to `handler.Validate()` so that `handler.updateIndex()` does not have this burden, however this would mean trusting what comes from the disk is not corrupt somehow. What do you think?
In my opinion, better safe than sorry.

Reply:
> this would mean trusting what comes from the disk is not corrupt somehow

I don't think we check any other chunks for corruption when reading from disk. I think if someone broke into your node's dbstore, nothing can be trusted....

Reply:
ok. Fixed!